File: generated_code/GeminiPro/Python/095_SplayTree.py
import sys

# It's good practice to set a higher recursion limit for tree-based structures,
# though this specific implementation is iterative and doesn't rely on deep recursion.
sys.setrecursionlimit(2000)

class SplayTree:
    """
    A complete, self-contained Splay Tree class that implements a
    dictionary-like set for storing unique integer keys.

    This implementation provides `insert`, `delete`, and `search` methods.
    A key characteristic of a Splay Tree is that any accessed node (through
    search, insertion, or deletion) is moved to the root of the tree using a
    series of rotations. This "splaying" operation provides amortized
    logarithmic time complexity for all operations, making the tree
    self-optimizing by keeping frequently accessed nodes near the top.
    """

    class _Node:
        """A private helper class representing a node in the Splay Tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
            
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent

        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Splays the node x to the root of the tree through a series of rotations.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:
                # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:
                # Zig-Zig case (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:
                # Zig-Zig case (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:
                # Zig-Zag case (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:
                # Zig-Zag case (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key and splays the accessed node to the root.

        If the key is found, the node with that key is splayed. If not found,
        the last non-null node accessed during the search is splayed.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found in the tree, False otherwise.
        """
        if not self.root:
            return False

        current_node = self.root
        last_node = self.root
        while current_node:
            last_node = current_node
            if key == current_node.key:
                self._splay(current_node)
                return True
            elif key < current_node.key:
                current_node = current_node.left
            else:
                current_node = current_node.right
        
        # Splay the last accessed node if the key was not found.
        self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a new integer key into the tree.

        If the key does not exist, it is inserted and the new node is splayed
        to the root. If the key already exists, the existing node is splayed
        to the root.

        Args:
            key (int): The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        # Use search to find the key. If it exists, search will splay it.
        if self.search(key):
            return

        # If key is not found, search() has splayed the closest node to the root.
        # Now, we insert the new key as the new root and restructure.
        new_node = self._Node(key)
        
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            self.root.left = None
            if new_node.left:
                new_node.left.parent = new_node
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            self.root.right = None
            if new_node.right:
                new_node.right.parent = new_node
        
        self.root.parent = new_node
        self.root = new_node

    def delete(self, key):
        """
        Deletes an integer key from the tree.

        First, the tree is searched for the key, which brings the node (or its
        closest neighbor) to the root. If the key is present at the root, it is
        removed, and its two subtrees are joined.

        Args:
            key (int): The integer key to delete.
        """
        # Splay the node to the root. If not found, splay the closest node.
        if not self.search(key):
            # Key is not in the tree, nothing to delete.
            return

        # After search, the node to delete is at the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree.
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect the left subtree to operate on it.
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree.
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this max_node. It will become the root of the left subtree.
            # After splaying, it will have no right child.
            # We use a temporary SplayTree instance to perform the splay
            # on the disconnected subtree.
            temp_tree = SplayTree()
            temp_tree.root = left_subtree
            temp_tree._splay(max_node)
            
            # The new root of the main tree is the splayed max_node.
            self.root = temp_tree.root
            
            # Attach the original right subtree to the new root.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/095_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree for an integer set.
 * A splay tree is a self-balancing binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * It performs basic operations such as insertion, look-up and removal in O(log n)
 * amortized time.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * This is a private static nested class as its implementation is tied to the SplayTree.
     */
    private static class Node {
        int key;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Public API Methods ---

    /**
     * Searches for a key in the tree.
     * If the key is found, it is splayed to the root. If not found, the last
     * accessed node on the search path is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree.
     * If the key is not already in the tree, it is inserted and becomes the new root.
     * If the key already exists, the existing node is splayed to the root.
     * This implementation maintains the set property (no duplicate keys).
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // Case 1: The tree is empty.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the node with the key, or its
        // would-be parent, to the root.
        root = splay(root, key);

        // Case 2: The key is already present. Do nothing as this is a set.
        if (root.key == key) {
            return;
        }

        // Case 3: The key is not present. Insert it as the new root.
        Node newNode = new Node(key);
        if (key < root.key) {
            // The new key is smaller than the root.
            // The new root's right child is the old root.
            // The new root's left child is the old root's left child.
            newNode.right = root;
            newNode.left = root.left;
            root.left = null; // The old root no longer has a left child.
        } else { // key > root.key
            // The new key is larger than the root.
            // The new root's left child is the old root.
            // The new root's right child is the old root's right child.
            newNode.left = root;
            newNode.right = root.right;
            root.right = null; // The old root no longer has a right child.
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree.
     * If the key is found, it is first splayed to the root and then removed.
     * The tree is then re-formed by joining the remaining subtrees.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the key to bring it to the root.
        root = splay(root, key);

        // If the key is not at the root after splaying, it wasn't in the tree.
        if (root.key != key) {
            return;
        }

        // Now, the node to delete is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay on the left subtree to bring its maximum element to its root.
            // Since `key` is guaranteed to be greater than any element in the
            // left subtree, splaying for `key` will bring the max element up.
            leftSubtree = splay(leftSubtree, key);

            // The new root of the left subtree (the max element) now has no right child.
            // We can attach the original right subtree there.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }

    // --- Private Helper Methods ---

    /**
     * Performs a right rotation on the given node y.
     *
     *      y           x
     *     / \         / \
     *    x   T3  ->  T1  y
     *   / \             / \
     *  T1  T2          T2  T3
     *
     * @param y The root of the subtree to rotate.
     * @return The new root of the rotated subtree (x).
     */
    private Node rotateRight(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Performs a left rotation on the given node x.
     *
     *    x             y
     *   / \           / \
     *  T1  y    ->   x   T3
     *     / \       / \
     *    T2  T3    T1  T2
     *
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree (y).
     */
    private Node rotateLeft(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splaying operation (top-down).
     * This operation moves the node with the given key (or the last accessed node on the
     * search path if the key is not found) to the root of the tree.
     *
     * @param node The root of the tree to splay.
     * @param key  The key to splay towards the root.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node node, int key) {
        if (node == null || node.key == key) {
            return node;
        }

        // Create a dummy node to act as the head of the Left and Right subtrees.
        // The key in the dummy node is irrelevant.
        Node dummy = new Node(0);
        Node leftTreeMax = dummy;
        Node rightTreeMin = dummy;
        Node current = node;

        while (true) {
            if (key < current.key) {
                if (current.left == null) break;
                // Zig-Zig (Left-Left case)
                if (key < current.left.key) {
                    current = rotateRight(current);
                    if (current.left == null) break;
                }
                // Link current node to the right tree (it's larger than the key)
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) break;
                // Zig-Zig (Right-Right case)
                if (key > current.right.key) {
                    current = rotateLeft(current);
                    if (current.right == null) break;
                }
                // Link current node to the left tree (it's smaller than the key)
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else { // key == current.key
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;   // Attach final left child
        rightTreeMin.left = current.right;  // Attach final right child
        current.left = dummy.right;         // Set new root's left child
        current.right = dummy.left;         // Set new root's right child

        return current;
    }
}